src/ast/passes/codegen_llvm.cpp

```diff
@@ -151,16 +151,98 @@
                                 StackType stack_type,
                                 const location &loc)
 {
-  Value *stackid = b_.CreateGetStackId(
-      ctx_, ident == "ustack", stack_type, loc);
+  if (!murmur_hash_2_func_)
+    murmur_hash_2_func_ = createMurmurHash2Func();
+
+  bool is_ustack = ident == "ustack";
+  auto uint64_size = sizeof(uint64_t);
+
+  AllocaInst *stackid = b_.CreateAllocaBPF(CreateUInt64(), "stackid");
+
+  Function *parent = b_.GetInsertBlock()->getParent();
+  BasicBlock *stack_scratch_failure = BasicBlock::Create(
+      module_->getContext(), "stack_scratch_failure", parent);
+  BasicBlock *merge_block = BasicBlock::Create(module_->getContext(),
+                                               "merge_block",
+                                               parent);
+
+  Value *stack_trace = b_.CreateGetStackScratchMap(stack_type,
+                                                   stack_scratch_failure,
+                                                   loc);
+  b_.CREATE_MEMSET(
+      stack_trace, b_.getInt8(0), uint64_size * stack_type.limit, 1);
+
+  BasicBlock *get_stack_success = BasicBlock::Create(module_->getContext(),
+                                                     "get_stack_success",
+                                                     parent);
+  BasicBlock *get_stack_fail = BasicBlock::Create(module_->getContext(),
+                                                  "get_stack_fail",
+                                                  parent);
+
+  Value *ptr_to_stack_trace = b_.CreatePointerCast(stack_trace,
+                                                   b_.GET_PTR_TY());
+  Value *stack_size = b_.CreateGetStack(
+      ctx_, is_ustack, ptr_to_stack_trace, stack_type, loc);
+  Value *condition = b_.CreateICmpSGE(stack_size, b_.getInt32(uint64_size));
+  b_.CreateCondBr(condition, get_stack_success, get_stack_fail);
+
+  b_.SetInsertPoint(get_stack_fail);
+  if (bpftrace_.debug_output_)
+    b_.CreateDebugOutput("Failed to get stack. Error: %d",
+                         std::vector<Value *>{ stack_size },
+                         loc);
+  b_.CreateStore(b_.getInt64(0), stackid);
+  b_.CreateBr(merge_block);
+  b_.SetInsertPoint(get_stack_success);
+
+  Value *stack_len = b_.CreateUDiv(stack_size, b_.getInt32(uint64_size));
+  // uint64 *stack, uint64 len, uint64 seed
+  AllocaInst *seed = b_.CreateAllocaBPF(CreateUInt64(), "seed");
+  // use the pid as the seed if we're getting user stacks
+  if (is_ustack) {
+    b_.CreateStore(b_.CreateGetPidTgid(loc), seed);
+  } else {
+    b_.CreateStore(b_.getInt64(1), seed);
```

Comment:
Can the seed be 0 for kernel? Cuz pid 1 is a real pid.

Reply:
I think the important thing here is that the seed is constant for all kstacks not that it's 1 or 0. @javierhonduco and I were debating if we even need a different seed based on pid. Maybe I should just leave it out entirely.

Reply:
I decide just to keep the seed constant for both kstack and ustack as it's mostly for randomization to prevent attacks, which isn't needed here. We also want to get the same stack_id if we have the same stack for two different pids.